crates/wasi/src/preview2/tcp.rs

```diff
@@ -103,23 +107,76 @@ impl HostInputStream for HostTcpSocketInner {
 
 const SOCKET_READY_SIZE: usize = 1024 * 1024 * 1024;
 
+pub(crate) struct TcpWriteStream {
+    stream: HostTcpSocketInner,
+    write_handle: Option<tokio::task::JoinHandle<anyhow::Result<()>>>,
+}
+
+impl TcpWriteStream {
+    pub(crate) fn new(stream: HostTcpSocketInner) -> Self {
+        Self {
+            stream,
+            write_handle: None,
+        }
+    }
+
+    /// Write `bytes` in a background task, remembering the task handle for use in a future call to
+    /// `write_ready`
+    fn background_write(&mut self, mut bytes: bytes::Bytes) {
+        assert!(self.write_handle.is_none());
+
+        let stream = self.stream.clone();
+        self.write_handle.replace(tokio::spawn(async move {
+            while !bytes.is_empty() {
+                stream.tcp_socket().writable().await?;
+                let n = stream.tcp_socket().try_write(&bytes)?;
+                let _ = bytes.split_to(n);
```

Comment:
Here I'd recommend using `AsyncWriteExt::write_all` since it'll bundle all this internally

Reply:
From a DM: we'll have to use `try_write` in a loop as tokio doesn't allow `write_all` with a non-mutable reference for `self`.